### ISSUE TITLE
`func` in `uretprobe` returns caller name

### ISSUE THREAD

**Issue by viktormalik** [bpftrace/bpftrace#2285]
Originally discovered by @xutao323.

Using the `func` builtin inside an `uretprobe` gives the name of the caller function. However, the documentation says:

> func - Name of the traced function

which suggests that the name of the callee (one that the `uretprobe` is attached to) should be returned.

The reason for this is that `func` reads the `ip` register which is at that time set to the return instruction in the caller function. This should be either fixed or at least marked in documentation.

### What reproduces the bug?

```
# cat test.c
#include <unistd.h>

int fun() { }

int main() {
    fun();
    sleep(1);  // this is necessary due to other bug, see issue #2284
}

# bpftrace -e 'uretprobe:test:fun { printf("%s\n", func); }'
Attaching 1 probe...
main
```

**Comment by ajor** [comment]
Some discussion here: https://lore.kernel.org/bpf/ZKuyKj8jrEPzWYMM@krava/T/#u

**Comment by viktormalik** [comment]
FYI, kernel now supports `get_func_ip` for uprobes (https://github.com/torvalds/linux/commit/a3c485a5d8d4)

**Comment by xutao323** [comment]
Got a chance to revisit. I tried recent Fedora `kernel-6.7.0-0.rc4.35.fc40` with #2645, and this issue can be resolved.

My steps:

1. Install latest Fedora [kernel](https://src.fedoraproject.org/rpms/kernel): the `get_func_ip` change is in newer kernels after 6.6, I used [kernel-6.7.0-0.rc4.35.fc40](https://koji.fedoraproject.org/koji/buildinfo?buildID=2327764)
2. Build #2645 locally: a small change to pass `get_func_ip` for both uprobe and kprobe on top of https://github.com/ajor/bpftrace2/commit/b8a14f44051ce4ece7e34b0062ca9be58e3ab9ea (test still fails, i don't know):
```
diff --git a/src/ast/passes/codegen_llvm.cpp b/src/ast/passes/codegen_llvm.cpp
index 92ea1fe7..db288ab9 100644
--- a/src/ast/passes/codegen_llvm.cpp
+++ b/src/ast/passes/codegen_llvm.cpp
@@ -278,7 +278,7 @@ void CodegenLLVM::visit(Builtin &builtin)
     // But only use the it for kretprobes, as the helper will fail for kprobes
     // placed within a function (not at the entry).
     auto probe_type = probetype(current_attach_point_->provider);
-    if (probe_type == ProbeType::kretprobe)
+    if (probe_type == ProbeType::kretprobe || probe_type == ProbeType::uretprobe)
     {
       expr_ = b_.CreateGetFuncIp(ctx_, builtin.loc);
     }
diff --git a/src/ast/passes/semantic_analyser.cpp b/src/ast/passes/semantic_analyser.cpp
index 4fdba408..e0530972 100644
--- a/src/ast/passes/semantic_analyser.cpp
+++ b/src/ast/passes/semantic_analyser.cpp
@@ -357,6 +357,7 @@ void SemanticAnalyser::visit(Builtin &builtin)
             << "The func builtin can not be used with '"
             << attach_point->provider << "' probes";
 
+      /*
       if (type == ProbeType::uretprobe)
       {
         // Disabled due to #2285
@@ -364,6 +365,7 @@ void SemanticAnalyser::visit(Builtin &builtin)
             << "The 'func' builtin is not available for uretprobes. Consider "
                "using 'probe' instead.";
       }
+       */
 
       if (type == ProbeType::kretprobe &&
           !bpftrace_.feature_->has_helper_get_func_ip())
diff --git a/tests/semantic_analyser.cpp b/tests/semantic_analyser.cpp
index 47f485ea..2b338d96 100644
--- a/tests/semantic_analyser.cpp
+++ b/tests/semantic_analyser.cpp
@@ -794,7 +794,7 @@ TEST(semantic_analyser, call_func)
   test("uprobe:/bin/sh:f { printf(\"%s\", func);  }", 0);
 
   test("kretprobe:f { func }", 0);
-  test("uretprobe:/bin/sh:f { func }", 1); // Disabled due to #2285
+  test("uretprobe:/bin/sh:f { func }", 0); // Disabled due to #2285
 
   MockBPFfeature nofeature(false);
   test(nofeature, "kretprobe:f { func }", 1);
```
3. The local built binary:
```
# ./bpftrace --info
System
  OS: Linux 6.7.0-0.rc4.35.fc40.aarch64 #1 SMP PREEMPT_DYNAMIC Mon Dec  4 15:48:19 UTC 2023
  Arch: aarch64

Build
  version: v0.18.0
  LLVM: 16.0.6
  unsafe probe: no
  bfd: yes
  libdw (DWARF support): yes

Kernel helpers
  probe_read: yes
  probe_read_str: yes
  probe_read_user: yes
  probe_read_user_str: yes
  probe_read_kernel: yes
  probe_read_kernel_str: yes
  get_current_cgroup_id: yes
  send_signal: yes
  override_return: no
  get_boot_ns: yes
  dpath: yes
  skboutput: yes
  get_func_ip: yes

Kernel features
  Instruction limit: 1000000
  Loop support: yes
  btf: yes
  module btf: yes
  map batch: yes
  uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): yes

Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes
  ringbuf: yes

Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes
  kprobe_multi: no
  raw_tp_special: yes
  iter: yes
```
4. Test `uretprobe func`: local built binary is correct now, while old version system binary still points to `ip` register
```
[root@fedora src]# ./bpftrace -e 'uretprobe:/root/test/test:fun { printf("%s\n", func); }'
Attaching 1 probe...
fun
^C

[root@fedora src]# bpftrace -e 'uretprobe:/root/test/test:fun { printf("%s\n", func); }'
Attaching 1 probe...
main
^C

```
cheers

**Comment by jordalgo** [comment]
Related issue (just adding for reference): https://github.com/iovisor/bpftrace/issues/68

**Comment by xutao323** [comment]
Tested `kretprobe func` in https://github.com/iovisor/bpftrace/issues/2285#issuecomment-1849465762 environment, #2645 is able to fix #68 as expected. So `func` works in both `kretprobe` and `uretprobe` now.

#2645 test:
```
[root@fedora src]# bpftrace -e 'kretprobe:vfs_read { printf("func: '%s'\n", func); exit(); }' 
Attaching 1 probe...
func: ksys_read
Lost 3859 events


[root@fedora src]# ./bpftrace -e 'kretprobe:vfs_read { printf("func: '%s'\n", func); exit(); }' 
Attaching 1 probe...
func: vfs_read
Lost 3849 events
```
#68 test:
```
[root@fedora src]# bpftrace -e 'kretprobe:vfs_* { @[func] = count(); }'
Attaching 76 probes...
^C

@[ksys_write]: 1
@[ksys_write]: 2
@[ksys_read]: 3
@[vfs_fstatat]: 22
@[__do_sys_newfstatat]: 23
@[vfs_fstat]: 72
@[vfs_getattr]: 73
@[do_open]: 144
@[ksys_read]: 179
[root@fedora src]# ./bpftrace -e 'kretprobe:vfs_* { @[func] = count(); }'
Attaching 76 probes...
^C

@[vfs_write]: 2
@[vfs_fstat]: 64
@[vfs_fstatat]: 65
@[vfs_getattr]: 72
@[vfs_getattr_nosec]: 73
@[vfs_open]: 144
@[vfs_read]: 254
```

**Comment by jordalgo** [comment]
@xutao323 Thanks for re-testing. We'll try to get that PR merged.
